### PR TITLE
ms compression shortint integer

### PR DESCRIPTION
**Pull request by mayeul-zama**

### CODE REVIEW

File: tfhe/src/shortint/server_key/tests/modulus_switch_compression.rs [hunk 1]
```diff
@@ -0,0 +1,39 @@
+use crate::shortint::keycache::KEY_CACHE;
+use crate::shortint::parameters::*;
+use crate::shortint::server_key::tests::parametrized_test::create_parametrized_test;
+use rand::Rng;
+
+const NB_TESTS: usize = 10;
```
Comment (IceTDrinker) [inline tfhe/src/shortint/server_key/tests/modulus_switch_compression.rs:6]: aren't those constants shared somehow ? Maybe they could be put in the tests/mod.rs wdyt ?

File: tfhe/src/shortint/server_key/tests/modulus_switch_compression.rs [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+    for _ in 0..NB_TESTS {
+        let clear = rng.gen::<u64>() % modulus_sup;
+
+        let ctxt = cks.unchecked_encrypt(clear);
+
+        let compressed_ct = sks.switch_modulus_and_compress(&ctxt);
+
+        let lookup_table = sks.generate_msg_lookup_table(|a| a + 1, ctxt.message_modulus);
+
+        let ressed_ct = sks.decompress_and_apply_lookup_table(&compressed_ct, &lookup_table);
```
Comment (IceTDrinker) [inline tfhe/src/shortint/server_key/tests/modulus_switch_compression.rs:33]: ressed_ct ?

File: tfhe/src/shortint/ciphertext/compressed_modulus_switched_ciphertext.rs [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+///
+/// // Can be serialized in a smaller buffer
+/// let compressed_ct = sks.switch_modulus_and_compress(&ctxt);
+///
+/// let decompressed_ct = sks.decompress(&compressed_ct);
+///
+/// let dec = cks.decrypt(&decompressed_ct);
+///
+/// assert_eq!(clear, dec);
+/// ```
```
Comment (IceTDrinker) [inline tfhe/src/shortint/ciphertext/compressed_modulus_switched_ciphertext.rs:33]: that docstring might be better on the compression and decompression method perhaps ?

or make the two methos point to this docstring using markdown to automatically scroll to the example section something like

```
see [`CompressedModulusSwitchedCiphertext `](`CompressedModulusSwitchedCiphertext#example`) for usage
```

File: tfhe/src/shortint/ciphertext/compressed_modulus_switched_ciphertext.rs [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+    compressed_modulus_switched_lwe_ciphertext: CompressedModulusSwitchedLweCiphertext<u64>,
+    degree: Degree,
+    message_modulus: MessageModulus,
+    carry_modulus: CarryModulus,
+    pbs_order: PBSOrder,
+}
+
+impl ServerKey {
+    // Packs a ciphertext to have a smaller serialization size
+    pub fn switch_modulus_and_compress(
```
Comment (IceTDrinker) [inline tfhe/src/shortint/ciphertext/compressed_modulus_switched_ciphertext.rs:45]: see comment above to either have the docstring here or redirect to the docstring above

File: tfhe/src/shortint/ciphertext/compressed_modulus_switched_ciphertext.rs [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+            carry_modulus: ct.carry_modulus,
+            pbs_order: ct.pbs_order,
+        }
+    }
+
+    // Unpacks a compressed ciphertext
+    // The degree from before the packing is conserved.
+    // This operation uses a PBS. Fot the same cost, it's possible to apply a lookup table by
+    // calling unpack_and_apply_lookup_table instead.
+    pub fn decompress(&self, compressed_ct: &CompressedModulusSwitchedCiphertext) -> Ciphertext {
```
Comment (IceTDrinker) [inline tfhe/src/shortint/ciphertext/compressed_modulus_switched_ciphertext.rs:87]: docstring thing, especially since you already have it written out basically :)

File: tfhe/src/shortint/ciphertext/compressed_modulus_switched_ciphertext.rs [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+
+        result
+    }
+
+    // Unpacks a compressed ciphertext
+    // This operation uses a PBS so we can apply a lookup table
+    // An identity lookup table may be applied to get the pre packing ciphertext with a nominal
+    // noise, however, it's better to call `unpack` for that because it conserves the degree
+    // instead of setting it to the  max of the lookup table
+    pub fn decompress_and_apply_lookup_table(
```
Comment (IceTDrinker) [inline tfhe/src/shortint/ciphertext/compressed_modulus_switched_ciphertext.rs:102]: docstring thing, especially since you already have it written out basically :)

File: tfhe/src/integer/server_key/radix/tests/modulus_switch_compression.rs [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+
+const NB_TESTS: usize = 10;
+
+create_parametrized_test!(modulus_switch_compression);
+create_parametrized_test!(modulus_switch_compression_signed);
+
+fn modulus_switch_compression<P>(param: P)
+where
+    P: Into<PBSParameters>,
+{
```
Comment (IceTDrinker) [inline tfhe/src/integer/server_key/radix/tests/modulus_switch_compression.rs:16]: did you check for the function executor pattern ? 

New tests are expected to use the function executor so that it's easy to test with different compute/hardware backends

File: tfhe/src/integer/server_key/radix/tests/modulus_switch_compression.rs [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+
+        let dec: u8 = cks.decrypt(&decompressed_ct);
+
+        assert_eq!(clear, dec);
+    }
+}
+
+fn modulus_switch_compression_signed<P>(param: P)
+where
+    P: Into<PBSParameters>,
```
Comment (IceTDrinker) [inline tfhe/src/integer/server_key/radix/tests/modulus_switch_compression.rs:40]: don't know if that's the pattern we use for test organization in integer, could you check with Thomas please ?

File: tfhe/src/integer/ciphertext/compressed_modulus_switched_ciphertext.rs [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+/// let dec = cks.decrypt(&decompressed_ct);
+///
+/// assert_eq!(clear, dec);
+/// ```
+pub type CompressedModulusSwitchedRadixCiphertext =
+    BaseRadixCiphertext<CompressedModulusSwitchedCiphertext>;
+
+impl ServerKey {
+    // Packs a ciphertext to have a smaller serialization size
+    pub fn switch_modulus_and_compress(
```
Comment (IceTDrinker) [inline tfhe/src/integer/ciphertext/compressed_modulus_switched_ciphertext.rs:38]: same remark as for shortint, your docstrings are mostly already written, if required, redirect to the docstring above and check (via `make doc`) that it renders ok, as here we have a type alias and not an "actual" type

File: tfhe/src/integer/ciphertext/compressed_modulus_switched_ciphertext.rs [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+            .blocks
+            .iter()
+            .map(|a| self.key.switch_modulus_and_compress(a))
+            .collect();
+
+        BaseRadixCiphertext { blocks }
+    }
+    // Unpacks a compressed ciphertext
+    // This operation costs a PBS
+    pub fn decompress(
```
Comment (IceTDrinker) [inline tfhe/src/integer/ciphertext/compressed_modulus_switched_ciphertext.rs:52]: docstring please :)

File: tfhe/src/integer/ciphertext/compressed_modulus_switched_ciphertext.rs [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+        &self,
+        compressed_ct: &CompressedModulusSwitchedRadixCiphertext,
+    ) -> RadixCiphertext {
+        let blocks = compressed_ct
+            .blocks
+            .iter()
+            .map(|a| self.key.decompress(a))
+            .collect();
+
+        BaseRadixCiphertext { blocks }
```
Comment (IceTDrinker) [inline tfhe/src/integer/ciphertext/compressed_modulus_switched_ciphertext.rs:62]: let's parallelize maybe as a PBS is not cheap ?

or have a decompress_parallelized

File: tfhe/src/integer/ciphertext/compressed_modulus_switched_ciphertext.rs [hunk 12]
```diff
@@ -0,0 +1,97 @@
...
+    }
+}
+
+pub type CompressedModulusSwitchedSignedRadixCiphertext =
+    BaseSignedRadixCiphertext<CompressedModulusSwitchedCiphertext>;
```
Comment (IceTDrinker) [inline tfhe/src/integer/ciphertext/compressed_modulus_switched_ciphertext.rs:67]: docstring for this type as well please

File: tfhe/src/integer/ciphertext/compressed_modulus_switched_ciphertext.rs [hunk 13]
```diff
@@ -0,0 +1,97 @@
...
+    // This operation costs a PBS
+    pub fn decompress_signed(
+        &self,
+        compressed_ct: &CompressedModulusSwitchedSignedRadixCiphertext,
+    ) -> SignedRadixCiphertext {
+        let blocks = compressed_ct
+            .blocks
+            .iter()
+            .map(|a| self.key.decompress(a))
+            .collect();
```
Comment (IceTDrinker) [inline tfhe/src/integer/ciphertext/compressed_modulus_switched_ciphertext.rs:93]: same remark for parallelization

File: tfhe/src/integer/ciphertext/compressed_modulus_switched_ciphertext.rs [hunk 14]
```diff
@@ -0,0 +1,97 @@
...
+
+        BaseRadixCiphertext { blocks }
+    }
+}
+
+pub type CompressedModulusSwitchedSignedRadixCiphertext =
+    BaseSignedRadixCiphertext<CompressedModulusSwitchedCiphertext>;
+
+impl ServerKey {
+    // Packs a ciphertext to have a smaller serialization size
```
Comment (IceTDrinker) [inline tfhe/src/integer/ciphertext/compressed_modulus_switched_ciphertext.rs:70]: docstring as it's mostly ready, redirect to relevant docstring for example if needed see shortint comments

File: tfhe/src/integer/ciphertext/compressed_modulus_switched_ciphertext.rs [hunk 15]
```diff
@@ -0,0 +1,97 @@
...
+    pub fn switch_modulus_and_compress(
+        &self,
+        ct: &RadixCiphertext,
+    ) -> CompressedModulusSwitchedRadixCiphertext {
+        let blocks = ct
+            .blocks
+            .iter()
+            .map(|a| self.key.switch_modulus_and_compress(a))
+            .collect();
```
Comment (IceTDrinker) [inline tfhe/src/integer/ciphertext/compressed_modulus_switched_ciphertext.rs:46]: parallelize as it may be costly in some cases due to the keyswitch

File: tfhe/src/integer/ciphertext/compressed_modulus_switched_ciphertext.rs [hunk 16]
```diff
@@ -0,0 +1,97 @@
...
+    // Packs a ciphertext to have a smaller serialization size
+    pub fn switch_modulus_and_compress_signed(
+        &self,
+        ct: &SignedRadixCiphertext,
+    ) -> CompressedModulusSwitchedSignedRadixCiphertext {
+        let blocks = ct
+            .blocks
+            .iter()
+            .map(|a| self.key.switch_modulus_and_compress(a))
+            .collect();
```
Comment (IceTDrinker) [inline tfhe/src/integer/ciphertext/compressed_modulus_switched_ciphertext.rs:79]: parallelization as KS may be costly done sequentially

can be a separate function

**Review by IceTDrinker — COMMENTED** [review]
Mostly details apart for tests where we now use the function executor to have tests that can be run easily on more hardware/backends

otherwise just missing some _parallelized variants of operations as in the KS_PBS case we keyswitch before compressing and we apply a PBS in all cases during decompression which can be costly done sequentially

thanks !

File: tfhe/src/shortint/server_key/tests/modulus_switch_compression.rs (hunk 1 above)
Comment (mayeul-zama) [reply tfhe/src/shortint/server_key/tests/modulus_switch_compression.rs]: Good ides
Moved

File: tfhe/src/shortint/server_key/tests/modulus_switch_compression.rs (hunk 2 above)
Comment (mayeul-zama) [reply tfhe/src/shortint/server_key/tests/modulus_switch_compression.rs]: Renamed `uncompressed_ct`

File: tfhe/src/shortint/ciphertext/compressed_modulus_switched_ciphertext.rs (hunk 3 above)
Comment (mayeul-zama) [reply tfhe/src/shortint/ciphertext/compressed_modulus_switched_ciphertext.rs:33]: Done

File: tfhe/src/shortint/ciphertext/compressed_modulus_switched_ciphertext.rs (hunk 4 above)
Comment (mayeul-zama) [reply tfhe/src/shortint/ciphertext/compressed_modulus_switched_ciphertext.rs:47]: Done

File: tfhe/src/shortint/ciphertext/compressed_modulus_switched_ciphertext.rs (hunk 5 above)
Comment (mayeul-zama) [reply tfhe/src/shortint/ciphertext/compressed_modulus_switched_ciphertext.rs:91]: Done

File: tfhe/src/integer/ciphertext/compressed_modulus_switched_ciphertext.rs (hunk 10 above)
Comment (mayeul-zama) [reply tfhe/src/integer/ciphertext/compressed_modulus_switched_ciphertext.rs]: Done

File: tfhe/src/integer/ciphertext/compressed_modulus_switched_ciphertext.rs (hunk 9 above)
Comment (mayeul-zama) [reply tfhe/src/integer/ciphertext/compressed_modulus_switched_ciphertext.rs]: Done

File: tfhe/src/integer/ciphertext/compressed_modulus_switched_ciphertext.rs (hunk 11 above)
Comment (mayeul-zama) [reply tfhe/src/integer/ciphertext/compressed_modulus_switched_ciphertext.rs]: Not sure if a non parallelized version is useful
Made parallelized

File: tfhe/src/integer/ciphertext/compressed_modulus_switched_ciphertext.rs (hunk 12 above)
Comment (mayeul-zama) [reply tfhe/src/integer/ciphertext/compressed_modulus_switched_ciphertext.rs]: Done

File: tfhe/src/integer/ciphertext/compressed_modulus_switched_ciphertext.rs (hunk 13 above)
Comment (mayeul-zama) [reply tfhe/src/integer/ciphertext/compressed_modulus_switched_ciphertext.rs]: Not sure if a non parallelized version is useful
Made parallelized

File: tfhe/src/integer/ciphertext/compressed_modulus_switched_ciphertext.rs (hunk 14 above)
Comment (mayeul-zama) [reply tfhe/src/integer/ciphertext/compressed_modulus_switched_ciphertext.rs]: Done

File: tfhe/src/integer/ciphertext/compressed_modulus_switched_ciphertext.rs (hunk 15 above)
Comment (mayeul-zama) [reply tfhe/src/integer/ciphertext/compressed_modulus_switched_ciphertext.rs]: Made parallelized

File: tfhe/src/integer/ciphertext/compressed_modulus_switched_ciphertext.rs (hunk 16 above)
Comment (mayeul-zama) [reply tfhe/src/integer/ciphertext/compressed_modulus_switched_ciphertext.rs]: Made parallelized

File: tfhe/src/integer/server_key/radix/tests/modulus_switch_compression.rs (hunk 7 above)
Comment (mayeul-zama) [reply tfhe/src/integer/server_key/radix/tests/modulus_switch_compression.rs:16]: I talked about it with Thomas
It's not useful as long as the operation is only supported on the CPU

File: tfhe/src/integer/ciphertext/compressed_modulus_switched_ciphertext.rs [hunk 17]
```diff
@@ -0,0 +1,132 @@
...
+    ) -> CompressedModulusSwitchedRadixCiphertext {
+        let blocks = ct
+            .blocks
+            .par_iter()
+            .map(|a| self.key.switch_modulus_and_compress(a))
+            .collect();
+
+        BaseRadixCiphertext { blocks }
+    }
+    /// Uncompresses a compressed ciphertext
```
Comment (IceTDrinker) [inline tfhe/src/integer/ciphertext/compressed_modulus_switched_ciphertext.rs:53]: decompress

File: tfhe/src/integer/ciphertext/compressed_modulus_switched_ciphertext.rs [hunk 18]
```diff
@@ -0,0 +1,132 @@
...
+///
+/// let dec = cks.decrypt(&decompressed_ct);
+///
+/// assert_eq!(clear, dec);
+/// ```
+pub type CompressedModulusSwitchedRadixCiphertext =
+    BaseRadixCiphertext<CompressedModulusSwitchedCiphertext>;
+
+impl ServerKey {
+    /// Compresses a ciphertext to have a smaller serialization size
```
Comment (IceTDrinker) [inline tfhe/src/integer/ciphertext/compressed_modulus_switched_ciphertext.rs:38]: I missed that the first time around but lets keep impl blocks for the ServerKey in the server_key/ directory, add a module if needed

File: tfhe/src/shortint/ciphertext/compressed_modulus_switched_ciphertext.rs [hunk 19]
```diff
@@ -0,0 +1,184 @@
...
+pub struct CompressedModulusSwitchedCiphertext {
+    compressed_modulus_switched_lwe_ciphertext: CompressedModulusSwitchedLweCiphertext<u64>,
+    degree: Degree,
+    message_modulus: MessageModulus,
+    carry_modulus: CarryModulus,
+    pbs_order: PBSOrder,
+}
+
+impl ServerKey {
+    /// Packs a ciphertext to have a smaller serialization size
```
Comment (IceTDrinker) [inline tfhe/src/shortint/ciphertext/compressed_modulus_switched_ciphertext.rs:44]: missed as well the first time around, let's keep impl blocks in server_key/ add a module if needed

**Review by IceTDrinker — COMMENTED** [review]
Two small things

Uncompress -> decompress

keep impl blocks for the server keys in server_key of their API level

there is a conflict to manage, otherwise looks good

File: tfhe/src/integer/server_key/radix/mod.rs [hunk 20]
```diff
@@ -1,6 +1,7 @@
 mod add;
 mod bitwise_op;
 mod comparison;
+mod modulus_switch_compression;
```
Comment (IceTDrinker) [inline tfhe/src/integer/server_key/radix/mod.rs:4]: the primitives are parallelized right ? so :/

let's move the module and have the primitive have the parallelized suffix, ok to not provide the sequential version, but we need consistency in the API, I think I mentionned the naming change

Comment (mayeul-zama) [reply tfhe/src/integer/server_key/radix/mod.rs:4]: Good catch

**Review by IceTDrinker — APPROVED** [review]
Thanks a lot !

File: tfhe/src/shortint/server_key/tests/modulus_switch_compression.rs [hunk 21]
```diff
@@ -0,0 +1,45 @@
+use super::NB_TESTS;
+use crate::shortint::keycache::KEY_CACHE;
+use crate::shortint::parameters::*;
+use crate::shortint::server_key::tests::parametrized_test::create_parametrized_test;
+use rand::Rng;
+
+create_parametrized_test!(shortint_modulus_switch_compression);
```
Comment (IceTDrinker) [inline tfhe/src/shortint/server_key/tests/modulus_switch_compression.rs:7]: this will fail with multi bit right ? so we need to use the version where we choose which parameters are tested

like

```rust
create_parametrized_test!({
    PARAM_SET1,
    PARAM_SET2,
});
```

File: tfhe/src/integer/server_key/radix_parallel/tests_unsigned/modulus_switch_compression.rs [hunk 22]
```diff
@@ -0,0 +1,42 @@
...
+fn modulus_switch_compression<P>(param: P)
+where
+    P: Into<PBSParameters>,
+{
+    let param = param.into();
+
+    // Modulus switch compression and multi bit PBS are currently not compatible
+    if let PBSParameters::MultiBitPBS(_) = &param {
+        return;
+    }
```
Comment (IceTDrinker) [inline tfhe/src/integer/server_key/radix_parallel/tests_unsigned/modulus_switch_compression.rs:21]: I don't like this it would look like the test was run,

prefer using the manual choice of parameters via

```rust
create_parameterized_test!({
    PARAM_SET1,
    PARAM_SET2,
});
```

then we'll revert the change in the macro once we have the fix for the multibit

File: tfhe/src/integer/server_key/radix_parallel/tests_signed/modulus_switch_compression.rs [hunk 23]
```diff
@@ -0,0 +1,42 @@
...
+fn modulus_switch_compression_signed<P>(param: P)
+where
+    P: Into<PBSParameters>,
+{
+    let param = param.into();
+
+    // Modulus switch compression and multi bit PBS are currently not compatible
+    if let PBSParameters::MultiBitPBS(_) = &param {
+        return;
+    }
```
Comment (IceTDrinker) [inline tfhe/src/integer/server_key/radix_parallel/tests_signed/modulus_switch_compression.rs:21]: prefer updating the macro with parameters as this is what is done elsewhere and avoids thinking tests are passing while they are not run

**Review by IceTDrinker — COMMENTED** [review]
Just one thing on how to skip the multi bit tests, let's not instantiate them rather than instantiating them and not returning early without executing

**Comment by IceTDrinker** [comment]
so https://github.com/zama-ai/tfhe-rs/blob/fb4b975c3464e0c70167b97643bd4f99b30e4f80/tfhe/src/integer/server_key/radix_parallel/tests_unsigned/mod.rs#L300

you should be able to choose the paramters depending on coverage in this way for integer, do you think it's worth updating one last time ?

**Comment by IceTDrinker** [comment]
I think it's worth doing the samll update to have the preferred style in integer in all our tests, sorry about all those small things 😅

**Review by IceTDrinker — APPROVED** [review]
thanks !

**Review by IceTDrinker — APPROVED** [review]
